txmix/onion_transport.py

```diff
@@ -0,0 +1,118 @@
+    # a unix domain socket listener for receiving inbound Tor onion service
+    # connections. Some security sandboxing environments might enforce using
+    # unix domain sockets.
+    onion_unix_socket = attr.ib(validator=attr.validators.instance_of(str), default=None)
+    onion_tcp_interface = attr.ib(validator=attr.validators.instance_of(str), default=None)
+    onion_tcp_port = attr.ib(validator=attr.validators.instance_of(int), default=None)
+
+    onion_host = attr.ib(validator=attr.validators.instance_of(str))
+    onion_port = attr.ib(validator=attr.validators.instance_of(int))
+    onion_key = attr.ib(validator=attr.validators.instance_of(bytes))
+
+    def register_protocol(self, protocol):
+        # XXX todo: assert that protocol provides the appropriate interface
+        self.protocol = protocol
+
+    def start(self):
+        """
+        make this transport begin listening on the specified interface and UDP port
+        interface must be an IP address
+        """
+        onion, onion_port = self.addr
+        assert onion_port > 1024
+        hs_strings = []
+        if self.onion_unix_socket is None:
+            local_socket_endpoint = "unix:%s" % self.onion_unix_socket
+        else:
+            local_socket_endpoint = "tcp:%s:%s" % (self.onion_tcp_interface, self.onion_tcp_port)
+        d = endpoints.connectProtocol(local_socket_endpoint, self)
+
+        def got_socket(result):
+            if self.onion_unix_socket is None:
+                hs_strings.append("%s %s:%s" % (self.onion_port, self.onion_tcp_interface, self.onion_tcp_port))
+            else:
+                hs_strings.append("%s unix:%s" % (self.onion_port, self.onion_unix_socket))
+            hs = txtorcon.torconfig.EphemeralHiddenService(hs_strings, key_blob_or_type=self.onion_key)
+            d2 = hs.add_to_tor(self.tor_control_protocol)
+            return d2
+        d.addCallback(got_socket)
+        return d
+
+    def send(self, addr, message):
+        """
+        send message to addr
+        where addr is a 2-tuple of type: (onion host, onion port)
+        """
+        tor_endpoint = endpoints.clientFromString("tor:%s:%s" % addr)
+        send_message_protocol = Protocol()
+
+        class OneShotSendProtocol(Protocol):
+            """
+            """
+        send_message_protocol = OneShotSendProtocol()
+        d = endpoints.connectProtocol(tor_endpoint, send_message_protocol)
+
+        def is_connected(protocol):
+            protocol.transport.write(message)
+            protocol.loseConnection()
+
+        d.addCallback(is_connected)
+        return d
+
+    # Protocol parent method overwriting
+
+    def dataReceived(self, data):
+        assert len(data) == self.sphinx_packet_size
+        self.protocol.received(data)
+
+    def connectionLost(self):
+        """
+        Called when the connection is shut down.
+        """
+        # XXX raise some kind of exception?
+        pass
```

Comment:
You can receive "reason" arg, and see if it's `ConnectionDone` (clean loss) or something else (which will be the error).